client/scala/src/main/scala/ai/verta/blobs/dataset/PathBlob.scala

```diff
@@ -0,0 +1,127 @@
+package ai.verta.blobs.dataset
+
+import ai.verta.swagger._public.modeldb.versioning.model._
+
+import java.security.{MessageDigest, DigestInputStream}
+import java.io.{File, FileInputStream}
+
+import scala.collection.mutable.HashMap
+import scala.util.{Failure, Success, Try}
+import scala.annotation.tailrec
+
+/** Captures metadata about files
+ *  @param paths list of filepaths or directory paths
+ *  @throws IOException - if any path is invalid (i.e non-existent)
+ *  @throws SecurityException - If a security manager exists and its SecurityManager.checkRead(java.lang.String) method denies read access to any file
+ */
+case class PathBlob(private val paths: List[String]) extends Dataset {
+  private val BufferSize = 8192
+
+  private val metadataList = paths.map(expanduser)
+  .flatMap((path: String) => processPath(new File(path)))
+  .map(metadata => metadata.path -> metadata)
+
+  protected var contents = HashMap(metadataList: _*)
+
+  override def equals(other: Any) = other match {
+    case other: PathBlob => contents.equals(other.contents)
+    case _ => false
+  }
+
+
+  /** Hash the file's content
+   *  From https://stackoverflow.com/questions/41642595/scala-file-hashing
+   *  @param path filepath
+   */
+  private def hash(file: File) = Try {
+    val buffer = new Array[Byte](BufferSize)
+    val messageDigest = MessageDigest.getInstance("MD5")
+
+    val dis = new DigestInputStream(
+      new FileInputStream(file),
+      messageDigest
+    )
+
+    try {
+      while (dis.read(buffer) != -1) {}
+    } finally {
+      dis.close()
+    }
+
+    // Convert to hexadecimal
+    messageDigest.digest.map("%02x".format(_)).mkString
+  }
+
+  /** Get the metadata of path.
+   *  @param file a file object, representing the path
+   *  @return a list of components of file under the path
+   */
+  private def processPath(file: File): List[FileMetadata] = dfs(List(file), List())
+
+  /** Tail-recursive DFS traversal to prevent stack overflow error
+   *  @param stack a stack containing the files/dirs to explore
+   *  @param acc accumulator list of file metadata to return
+   */
+  @tailrec private def dfs(stack: List[File], acc: List[FileMetadata]): List[FileMetadata] = {
+    if (stack.isEmpty) acc
+    else if (stack.head.isDirectory) {
+      val dir = stack.head
+      dfs(dir.listFiles().toList ::: stack.drop(1), acc)
```

Comment:
```suggestion
      dfs(dir.listFiles().toList ::: stack.tail, acc)
```

Reply:
oh that was dumb. should be fixed already.